File: animated-drawable/src/main/java/com/facebook/fresco/animation/frame/DropFramesFrameScheduler.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.fresco.animation.frame;

import com.facebook.common.internal.VisibleForTesting;
import com.facebook.fresco.animation.backend.AnimationInformation;

/**
 * Frame scheduler that maps time values to frames.
 */
public class DropFramesFrameScheduler implements FrameScheduler {

  private static final int UNSET = -1;

  private final AnimationInformation mAnimationInformation;

  private long mLoopDurationMs = UNSET;

  public DropFramesFrameScheduler(AnimationInformation animationInformation) {
    mAnimationInformation = animationInformation;
  }

  @Override
  public int getFrameNumberToRender(long animationTimeMs, long lastFrameTimeMs) {
    if (!isInfiniteAnimation()) {
      long loopCount = animationTimeMs / getLoopDurationMs();
      if (loopCount >= mAnimationInformation.getLoopCount()) {
        return FRAME_NUMBER_DONE;
      }
    }
    long timeInCurrentLoopMs = animationTimeMs % getLoopDurationMs();
    return getFrameNumberWithinLoop(timeInCurrentLoopMs);
  }

  @Override
  public long getLoopDurationMs() {
    if (mLoopDurationMs != UNSET) {
      return mLoopDurationMs;
    }
    mLoopDurationMs = 0;
    int frameCount = mAnimationInformation.getFrameCount();
    for (int i = 0; i < frameCount; i++) {
      mLoopDurationMs += mAnimationInformation.getFrameDurationMs(i);
    }
    return mLoopDurationMs;
  }

  @Override
  public long getTargetRenderTimeMs(int frameNumber) {
    long targetRenderTimeMs = 0;
    for (int i = 0; i < frameNumber; i++) {
      targetRenderTimeMs += mAnimationInformation.getFrameDurationMs(frameNumber);
    }
    return targetRenderTimeMs;
  }

  @Override
  public long getTargetRenderTimeForNextFrameMs(long animationTimeMs) {
    long loopDurationMs = getLoopDurationMs();
    // Sanity check.
    if (loopDurationMs == 0) {
      return NO_NEXT_TARGET_RENDER_TIME;
    }
    if (!isInfiniteAnimation()) {
      long loopCount = animationTimeMs / getLoopDurationMs();
      if (loopCount >= mAnimationInformation.getLoopCount()) {
        return NO_NEXT_TARGET_RENDER_TIME;
      }
    }
    // The animation time in the current loop
    long timePassedInCurrentLoopMs = animationTimeMs % loopDurationMs;
    // The animation time in the current loop for the next frame
    long timeOfNextFrameInLoopMs = 0;

    int frameCount = mAnimationInformation.getFrameCount();
    for (int i = 0; i < frameCount && timeOfNextFrameInLoopMs <= timePassedInCurrentLoopMs; i++) {
      timeOfNextFrameInLoopMs += mAnimationInformation.getFrameDurationMs(i);
    }

    // Difference between current time in loop and next frame in loop
    long timeUntilNextFrameInLoopMs = timeOfNextFrameInLoopMs - timePassedInCurrentLoopMs;
    // Add the difference to the current animation time
    return animationTimeMs + timeUntilNextFrameInLoopMs;
  }

  @Override
  public boolean isInfiniteAnimation() {
    return mAnimationInformation.getLoopCount() == AnimationInformation.LOOP_COUNT_INFINITE;
  }

  @VisibleForTesting
  int getFrameNumberWithinLoop(long timeInCurrentLoopMs) {
    int frame = 0;
    long currentDuration = 0;
    do {
      currentDuration += mAnimationInformation.getFrameDurationMs(frame);
      frame++;
    } while (timeInCurrentLoopMs >= currentDuration);
    return frame - 1;
  }
}

File: animated-drawable/src/main/java/com/facebook/fresco/animation/frame/DropFramesFrameScheduler.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.animation.frame

import androidx.annotation.VisibleForTesting
import com.facebook.fresco.animation.backend.AnimationInformation

/** Frame scheduler that maps time values to frames. */
class DropFramesFrameScheduler(private val animationInformation: AnimationInformation) :
    FrameScheduler {

  private var _loopDurationMs = UNSET.toLong()

  override fun getFrameNumberToRender(animationTimeMs: Long, lastFrameTimeMs: Long): Int {
    val loopDurationMs = this.loopDurationMs
    if (loopDurationMs == 0L) {
      return getFrameNumberWithinLoop(0)
    }
    if (!isInfiniteAnimation) {
      val loopCount = animationTimeMs / loopDurationMs
      if (loopCount >= animationInformation.loopCount) {
        return FrameScheduler.FRAME_NUMBER_DONE
      }
    }
    val timeInCurrentLoopMs = animationTimeMs % loopDurationMs
    return getFrameNumberWithinLoop(timeInCurrentLoopMs)
  }

  override fun getLoopDurationMs(): Long {
    if (_loopDurationMs != UNSET.toLong()) {
      return _loopDurationMs
    }
    _loopDurationMs = 0
    val frameCount = animationInformation.frameCount
    for (i in 0 until frameCount) {
      _loopDurationMs += animationInformation.getFrameDurationMs(i).toLong()
    }
    return _loopDurationMs
  }

  override fun getTargetRenderTimeMs(frameNumber: Int): Long {
    var targetRenderTimeMs = 0L
    for (i in 0 until frameNumber) {
      targetRenderTimeMs += animationInformation.getFrameDurationMs(frameNumber).toLong()
    }
    return targetRenderTimeMs
  }

  override fun getTargetRenderTimeForNextFrameMs(animationTimeMs: Long): Long {
    val loopDurationMs = this.loopDurationMs
    // Sanity check.
    if (loopDurationMs == 0L) {
      return FrameScheduler.NO_NEXT_TARGET_RENDER_TIME.toLong()
    }
    if (!isInfiniteAnimation) {
      val loopCount = animationTimeMs / loopDurationMs
      if (loopCount >= animationInformation.loopCount) {
        return FrameScheduler.NO_NEXT_TARGET_RENDER_TIME.toLong()
      }
    }
    // The animation time in the current loop
    val timePassedInCurrentLoopMs = animationTimeMs % loopDurationMs
    // The animation time in the current loop for the next frame
    var timeOfNextFrameInLoopMs = 0L
    val frameCount = animationInformation.frameCount
    var i = 0
    while (i < frameCount && timeOfNextFrameInLoopMs <= timePassedInCurrentLoopMs) {
      timeOfNextFrameInLoopMs += animationInformation.getFrameDurationMs(i).toLong()
      i++
    }

    // Difference between current time in loop and next frame in loop
    val timeUntilNextFrameInLoopMs = timeOfNextFrameInLoopMs - timePassedInCurrentLoopMs
    // Add the difference to the current animation time
    return animationTimeMs + timeUntilNextFrameInLoopMs
  }

  override fun isInfiniteAnimation(): Boolean =
      animationInformation.loopCount == AnimationInformation.LOOP_COUNT_INFINITE

  @VisibleForTesting
  fun getFrameNumberWithinLoop(timeInCurrentLoopMs: Long): Int {
    var frame = 0
    var currentDuration = 0L
    do {
      currentDuration += animationInformation.getFrameDurationMs(frame).toLong()
      frame++
    } while (timeInCurrentLoopMs >= currentDuration)
    return frame - 1
  }

  companion object {
    private const val UNSET = -1
  }
}
